### DISCUSSION TITLE
Soroban Bitcoin Price Oracle

### DISCUSSION THREAD

**Discussion by snowstorm134** [stellar/sorobounty-spectacular#29] [General]
Soroban Bitcoin Price Oracle - bitcoin price oracle built on Soroban Futurenet

Codebase: https://github.com/snowstorm134/SorobanCryptoOracle
Tutorial: https://dev.to/user1122/soroban-bitcoin-price-oracle-tutorial-3ldk
Demo: https://soroban-bitcoin-oracle.netlify.app/
Video: [Link](https://www.youtube.com/watch?v=YEHb36HEUyc)

**Comment by briwylde08** [comment]
Hello again! We’ll review your Sorobounty Spectacular submission and be in contact with you within the next week. 🎉

**Comment by Julian-dev28** [comment]
Hey @snowstorm134 

Awesome submission. Really great implementation of a much needed example for Soroban. 

Is there a reason why you are using 
```
cargo install_soroban 
```

to install the soroban alias vs our regular setup?

**Reply by Julian-dev28** [reply]
A few things:

re: soroban-cli installation
The current soroban-cli can generate bindings and users wouldn't have to install a special version of it.

Also you would have to add a `.cargo` directory and `config.toml` to tell cargo what to do when running ``install_soroban`` which is missing from the repo.
[reference]( https://github.com/stellar/soroban-example-dapp/blob/main/.cargo/config.toml)

I recommend removing this step and just using the standard soroban-cli


re: Initialize Social Network Contract
Please take a look at the initialize piece of the dev.to article as there may be some content that does not belong here: 

re: cron script
running `npm cron-script.js` as specified in the readme resulted in an error for the command. did you mean to tell the users to run `node cron-script.js` instead?

**Comment by snowstorm134** [comment]
> Hey @snowstorm134
> 
> Awesome submission. Really great implementation of a much needed example for Soroban.

Hi @Julian-dev28 . Thank you.

----




> Is there a reason why you are using
> 
> ```
> cargo install_soroban 
> ```
> 
> to install the soroban alias vs our regular setup?


> I recommend removing this step and just using the standard soroban-cli

Yes, you are right.
This is an unnecessary step. The fact is that I followed one guide from the documentation and there was this step. But then I realized that it can be done using standard `soroban-cli`

So I removed this step.

-----

> re: Initialize Social Network Contract
> Please take a look at the initialize piece of the dev.to article as there may be some content that does not belong here:

Thank you. I fixed it.

----

> re: cron script
> running `npm cron-script.js` as specified in the readme resulted in an error for the command. did you mean to tell the users to run `node cron-script.js` instead?

Yes, you are right. That's my fault.
The correct command is `node cron-script.js`.
I already fixed it. 

---

Thanks for the feedback.

**Reply by Julian-dev28** [reply]
No problem

A few more things:

1. are you able to add a quick note about how a user may have to run
```
chmod +x initialize.sh
```
for the script to work?

2. Is the user supposed to change these values in `initialize.sh`?

```
  --caller GCSXUXZSA2VEXN5VGOWE5ODAJLC575JCMWRJ4FFRDWSTRCJYQK4ML6V3 \
  --pair_name BTC_USDT \
  --epoch_interval 600 \
  --relayer GCSXUXZSA2VEXN5VGOWE5ODAJLC575JCMWRJ4FFRDWSTRCJYQK4ML6V3
```

**Reply by snowstorm134** [reply]
> 1. are you able to add a quick note about how a user may have to run
> 
> ```
> chmod +x initialize.sh
> ```

Okay. Will add this step in tutorial. This command gives execute permission to run this bash script.

**Reply by snowstorm134** [reply]
> 2\. Is the user supposed to change these values in `initialize.sh`?
> 
> ```
>   --caller GCSXUXZSA2VEXN5VGOWE5ODAJLC575JCMWRJ4FFRDWSTRCJYQK4ML6V3 \
>   --pair_name BTC_USDT \
>   --epoch_interval 600 \
>   --relayer GCSXUXZSA2VEXN5VGOWE5ODAJLC575JCMWRJ4FFRDWSTRCJYQK4ML6V3
> ```

The `initialize` function can only be called once. By calling `initialize`, the user specifies:
- `caller` - the address that will become the owner of the contract;
- `pair_name` - pair name. In fact, this is just a text, user can leave it like this;
- `epoch_interval ` - frequency (in seconds) of price updates;
- `relayer` - address of the wallet that will update the price (in the backend, in the cron task). It is better to create a separate wallet for this (with test tokens);


But some data (`relayer` and  `epoch_inteval`) can be changed later, but ONLY by the contract owner (the address specified as the `caller` parameter when calling the `initiliaze`).
- `update_pair_epoch_interval` - change epoch_interval
- `update_relayer_address` - change relayer address

**Reply by Julian-dev28** [reply]
Amazing

Thank you for the update.

Is there anything that a user would need to do to square a way the front end? ie install any libraries etc?

Everything is looking good on my end besides the actual price card.

![Screenshot 2023-09-05 at 11 37 03 AM](https://github.com/stellar/sorobounty-spectacular/assets/73849597/9c2ad907-e617-4af5-80ee-d7b27d280e24)


The data returns fine via soroban-cli so thats good:
```
soroban contract invoke \
--wasm target/wasm32-unknown-unknown/release/soroban_oracle_contract.wasm \
--id CASQCDKTBKNDP2LSYQLHLPKTDKRRPTK3Y2DEA2JKM7H3IQNIBE32T5NY \
--source <secret_key> \
--network futurenet \
-- \
get_pair_info
```

```
{"create_time":1693938377,"epoch_interval":600,"last_epoch":1,"pair_name":"BTC_USDT","relayer":"GCSXUXZSA2VEXN5VGOWE5ODAJLC575JCMWRJ4FFRDWSTRCJYQK4ML6V3"}
```

**Reply by snowstorm134** [reply]
I'm not sure, but let's try.

- go to dir `.soroban/oracle-contract/dist/esm/` 
- open file `index.js`,
- fing ALL `export async function` and replace:

> about 9 places need to be replaced. Somewhere there may already be the right value

the code here
```
parseResultXdr: (xdr) => {
            XXXXXXXXXXXXXXXXX
        },
```

with 
```
parseResultXdr: (xdr) => {
            return scValStrToJs(xdr);
        },
```

**Reply by snowstorm134** [reply]
let me know if you managed to find this file

**Reply by Julian-dev28** [reply]
Nicely done!
![Screenshot 2023-09-05 at 12 07 59 PM](https://github.com/stellar/sorobounty-spectacular/assets/73849597/c00573bb-e825-479a-90d2-7cda45bae398)


Is there away to fix this without manually changing these values?

Also, Should this calculator on the deposit page display BTC -> USDT instead of  USDT - > BTC
![Screenshot 2023-09-05 at 12 11 50 PM](https://github.com/stellar/sorobounty-spectacular/assets/73849597/8aa5a85a-7646-499b-b780-f789fb54b6b2)

**Reply by snowstorm134** [reply]
> Is there away to fix this without manually changing these values?

I'm not sure. It took me a long time to find this solution.

---

> Also, Should this calculator on the deposit page display BTC -> USDT instead of USDT - > BTC

Oh, it seems that the repository does not contain the final version of the frontend.
You can look at the live version of the app for now, there is an updated interface. It's more clear there. All values are shown in both USD and BTC.
But the logic remains the same: The user indicates the amount in USD and the amount converted to BTC is shown a little lower. Thus, I wanted to more clearly show the sense of the oracle.

I will update the repository soon. It is okay?

**Reply by Julian-dev28** [reply]
> I'm not sure. It took me a long time to find this solution.

Okay. Can you please add a section that shows users how to do this? 

In the meantime, I will make a note of this issue and see what other teams have to say about it.

>I will update the repository soon. It is okay?

Yes. That's all good. Just ping me when it is ready.

**Reply by snowstorm134** [reply]
> Okay. Can you please add a section that shows users how to do this?

Okay. I will add this info to the guide as soon as possible (today or tomorrow)

**Comment by Julian-dev28** [comment]
@snowstorm134 Hey there!

Just saw the update to the read me and I have a suggestion on the copy:

Are we able to change this statement:
> The fact is that at the moment, there is a certain error due to which the default created file generates an error in the frontend.

to:

Soroban-tooling is still in development, and the team is working to improve generated bindings that may not fully integrate with some frontends at this time.

In this project we fix this by changing the following code...(code snippet)

**Reply by snowstorm134** [reply]
Hi.

Yes, the text you suggested looks better.
Thank you.

I just changed it.

**Reply by Julian-dev28** [reply]
Looks great!


Did you plan on following up on this:

> Oh, it seems that the repository does not contain the final version of the frontend.
You can look at the live version of the app for now, there is an updated interface. It's more clear there. All values are shown in both USD and BTC.
But the logic remains the same: The user indicates the amount in USD and the amount converted to BTC is shown a little lower. Thus, I wanted to more clearly show the sense of the oracle.
I will update the repository soon. It is okay?


Or did you want to finalize the submission as is?

**Reply by snowstorm134** [reply]
Hi

> Did you plan on following up on this:

It will be done within one hour.

**Reply by Julian-dev28** [reply]
Amazing. 

Thanks for putting everything together and working with us!

**Reply by snowstorm134** [reply]
I just updated the frontend.
Let me know if anything else needs to be done.

**Reply by Julian-dev28** [reply]
Amazing

Got slightly hung up specifying relayers here again: 
```
echo "Initialize the ORACLE contract"
soroban contract invoke \
  $ARGS \
  --id "$ORACLE_ID" \
  -- \
  initialize \
  --caller GCSXUXZSA2VEXN5VGOWE5ODAJLC575JCMWRJ4FFRDWSTRCJYQK4ML6V3 \
  --pair_name BTC_USDT \
  --epoch_interval 600 \
  --relayer GCSXUXZSA2VEXN5VGOWE5ODAJLC575JCMWRJ4FFRDWSTRCJYQK4ML6V3
echo "Done"
```

So its a good idea to update the readme and make users aware of this. 
Other than that, I think this submission is great and am happy to have reviewed it!

Im going to wrap the review process up and Bri should be getting back to you shortly after.



Are you open to further collaboration? It would be great to implement this project into a challenge in the Soroban Dapps Challenge!

**Reply by snowstorm134** [reply]
Hi. 
Sorry for late response.

---

> So its a good idea to update the readme and make users aware of this.

I added this.

---

> Are you open to further collaboration? It would be great to implement this project into a challenge in the Soroban Dapps Challenge!

Yes, I am.
I think we can try to do this. Let me know more details

**Reply by Julian-dev28** [reply]
Absolutely

It would be great to implement this into the Soroban Dapps Challenge

So far, I've updated the dapp to work with preview release 11 here  : https://github.com/stellar/soroban-dapps-challenge/tree/update-crypto-oracle

If you have time to collaborate, it would be great to get your review on the changes so far:
- added multi wallet support
- updated to soroban sdk rc-2
- uses updated typescript bindings


Feel free to reach me on discord `@TheGodHand#0021` if that works better for you.

**Comment by briwylde08** [comment]
Hi @snowstorm134! Thanks so much for your work on this! [🎉](https://emojipedia.org/party-popper/)

Given our guidelines, your submission may qualify for an award of $1,500 worth of XLM. To receive that award, please fill out [this form](https://stellarform.typeform.com/to/HBjEyCJt) completely and accurately.

Thanks again! Please let us know if you have any questions.

Best,
Bri

**Reply by snowstorm134** [reply]
Hi.
Thanks. I will do this in the next few days.

**Comment by briwylde08** [comment]
Hi @snowstorm134! Thank you for providing the requested information. You’ve passed KYC and we’ve logged your award payment of $3,000 worth of XLM (this combines this submission plus https://github.com/stellar/sorobounty-spectacular/discussions/28). This should be sent to your wallet within the next week.

Please let me know if you have any questions! 

Best,
Bri

**Reply by snowstorm134** [reply]
Hi @briwylde08 .

Thanks ! 

I received the award.
